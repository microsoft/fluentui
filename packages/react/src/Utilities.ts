import './version';
export * from '@fluentui/utilities';
