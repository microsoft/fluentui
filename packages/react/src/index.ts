export * from './ActivityItem';
export * from './Autofill';
export * from './Announced';
export * from './Breadcrumb';
export * from './Button';
export * from './ButtonGrid';
export * from './Calendar';
export * from './Callout';
export * from './Check';
export * from './Checkbox';
export * from './ChoiceGroup';
// export * from './ChoiceGroupOption'; // exported by ChoiceGroup
export * from './Coachmark';
export * from './Color';
export * from './ColorPicker';
export * from './ComboBox';
export * from './CommandBar';
export * from './ContextualMenu';
export * from './DatePicker';
export * from './DateTimeUtilities';
export * from './DetailsList';
export * from './Dialog';
export * from './Divider';
export * from './DocumentCard';
export * from './DragDrop';
export * from './Dropdown';
export * from './ExtendedPicker';
export * from './Fabric';
export * from './Facepile';
export * from './FloatingPicker';
export * from './FocusTrapZone';
export * from './FocusZone';
export * from './Grid';
export * from './GroupedList';
export * from './HoverCard';
export * from './Icon';
export * from './Icons';
export * from './Image';
export * from './Keytips';
export * from './Keytip';
export * from './KeytipData';
export * from './KeytipLayer';
export * from './Label';
export * from './Layer';
export * from './Link';
export * from './List';
export * from './MarqueeSelection';
export * from './MessageBar';
export * from './Modal';
export * from './Nav';
export * from './OverflowSet';
export * from './Overlay';
export * from './Panel';
export * from './Persona';
export * from './PersonaCoin';
// export * from './PersonaPresence'; (Exported as part of Persona)
export * from './Pickers';
export * from './Pivot';
export * from './Popup';
export * from './Positioning';
export * from './PositioningContainer';
export * from './ProgressIndicator';
export * from './Rating';
export * from './ResizeGroup';
export * from './ResponsiveMode';
export * from './ScrollablePane';
export * from './SearchBox';
export * from './SelectableOption';
export * from './SelectedItemsList';
export * from './Selection';
export * from './Separator';
export * from './Shimmer';
export * from './ShimmeredDetailsList';
export * from './Slider';
export * from './SpinButton';
export * from './Spinner';
export * from './Stack';
export * from './Sticky';
export * from './Styling';
export * from './SwatchColorPicker';
export * from './TeachingBubble';
export * from './Text';
export * from './TextField';
export * from './Theme';
export * from './ThemeGenerator';
export * from './TimePicker';
export * from './Toggle';
export * from './Tooltip';
export * from './Utilities';
export * from './WeeklyDayPicker';
export * from './WindowProvider';

import './version';
