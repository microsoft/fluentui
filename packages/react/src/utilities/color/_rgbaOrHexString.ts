import { MAX_COLOR_ALPHA } from './consts';

/**
 * @internal
 * Get a CSS color string from some color components.
 * If `a` is specified and not 100, returns an `rgba()` string.
 * Otherwise returns `hex` prefixed with #.
 */
export function _rgbaOrHexString(r: number, g: number, b: number, a: number | undefined, hex: string): string {
  return a === MAX_COLOR_ALPHA || typeof a !== 'number' ? `#${hex}` : `rgba(${r}, ${g}, ${b}, ${a / MAX_COLOR_ALPHA})`;
}
