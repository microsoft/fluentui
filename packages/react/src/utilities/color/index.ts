export * from './colors';
export * from './shades';
export * from './consts';
export * from './interfaces';
export * from './cssColor';
export * from './rgb2hex';
export * from './clamp';
export * from './hsl2rgb';
export * from './hsl2hsv';
export * from './hsv2rgb';
export * from './hsv2hex';
export * from './rgb2hsv';
export * from './hsv2hsl';
export * from './getColorFromString';
export * from './getColorFromRGBA';
export * from './getColorFromHSV';
export * from './getFullColorString';
export * from './updateSV';
export * from './updateH';
export * from './updateRGB';
export * from './getColorFromString';
export * from './updateA';
export * from './updateT';
export * from './correctRGB';
export * from './correctHSV';
