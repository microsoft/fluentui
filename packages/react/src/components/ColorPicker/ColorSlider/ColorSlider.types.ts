import * as React from 'react';
import { ITheme, IStyle } from '../../../Styling';
import { IRefObject, IStyleFunctionOrObject } from '../../../Utilities';

/**
 * {@docCategory ColorPicker}
 */
export interface IColorSlider {
  /** Current value of the slider. */
  value: number;
}

/**
 * {@docCategory ColorPicker}
 */
export interface IColorSliderProps {
  /**
   * Gets the component ref.
   */
  componentRef?: IRefObject<IColorSlider>;

  /**
   * Minimum value of the slider.
   * @deprecated Will always be 0
   */
  minValue?: number;

  /**
   * Maximum value of the slider.
   * @deprecated Will be 100 for alpha or transparency sliders, or 359 for hue sliders.
   */
  maxValue?: number;

  /**
   * Current value of the slider.
   */
  value?: number;

  /**
   * Label of the ColorSlider for the benefit of screen reader users.
   */
  ariaLabel?: string;

  /**
   * Type of slider to display.
   * @defaultvalue 'hue'
   */
  type?: 'hue' | 'alpha' | 'transparency';

  /**
   * If true, the slider represents an alpha slider and will display a gray checkered pattern
   * in the background. Otherwise, the slider represents a hue slider.
   * @defaultvalue false
   * @deprecated Use `type`
   */
  isAlpha?: boolean;

  /**
   * Hex color to use when rendering an alpha or transparency slider's overlay, *without* the `#`.
   */
  overlayColor?: string;

  /**
   * CSS-compatible string for the color of the thumb element.
   * @deprecated Not used. Use `styles.sliderThumb` instead.
   */
  thumbColor?: string;

  /**
   * Custom style for the overlay element.
   * @deprecated Use `overlayColor` instead
   */
  overlayStyle?: React.CSSProperties;

  /**
   * Callback issued when the value changes.
   */
  onChange?: (event: React.MouseEvent | React.KeyboardEvent, newValue?: number) => void;

  /**
   * Additional CSS class(es) to apply to the ColorSlider.
   */
  className?: string;

  /**
   * Theme (provided through customization).
   */
  theme?: ITheme;

  /**
   * Call to provide customized styling that will layer on top of the variant rules.
   */
  styles?: IStyleFunctionOrObject<IColorSliderStyleProps, IColorSliderStyles>;
}

/**
 * {@docCategory ColorPicker}
 */
export type IColorSliderStyleProps = Required<Pick<IColorSliderProps, 'theme'>> &
  Pick<IColorSliderProps, 'className' | 'type'> & {
    /**
     * @deprecated Use `type`
     */
    isAlpha?: boolean;
  };

/**
 * {@docCategory ColorPicker}
 */
export interface IColorSliderStyles {
  /**
   * Style set for the root element.
   */
  root?: IStyle;

  /**
   * Style set for the draggable thumb element.
   */
  sliderThumb?: IStyle;

  /**
   * Style set for the overlay element.
   */
  sliderOverlay?: IStyle;
}
