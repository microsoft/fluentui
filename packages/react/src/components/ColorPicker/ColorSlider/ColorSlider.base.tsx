import * as React from 'react';
import {
  classNamesFunction,
  initializeComponentRef,
  on,
  KeyCodes,
  getWindow,
  warnDeprecations,
  warn,
} from '../../../Utilities';
import { IColorSliderProps, IColorSliderStyleProps, IColorSliderStyles, IColorSlider } from './ColorSlider.types';
import { clamp } from '../../../utilities/color/clamp';
import { MAX_COLOR_HUE, MAX_COLOR_ALPHA } from '../../../utilities/color/consts';

const getClassNames = classNamesFunction<IColorSliderStyleProps, IColorSliderStyles>();

export interface IColorSliderState {
  currentValue: number;
}

/**
 * {@docCategory ColorPicker}
 */
export class ColorSliderBase extends React.Component<IColorSliderProps, IColorSliderState> implements IColorSlider {
  public static defaultProps: Partial<IColorSliderProps> = {
    value: 0,
  };

  private _disposables: (() => void)[] = [];
  private _root = React.createRef<HTMLDivElement>();

  constructor(props: IColorSliderProps) {
    super(props);

    initializeComponentRef(this);

    warnDeprecations('ColorSlider', props, {
      thumbColor: 'styles.sliderThumb',
      overlayStyle: 'overlayColor',
      isAlpha: 'type',
      maxValue: 'type',
      minValue: 'type',
    });
    // eslint-disable-next-line deprecation/deprecation
    if (this._type !== 'hue' && !(props.overlayColor || props.overlayStyle)) {
      warn(`ColorSlider: 'overlayColor' is required when 'type' is "alpha" or "transparency"`);
    }

    this.state = {
      currentValue: props.value || 0,
    };
  }

  public get value(): number {
    return this.state.currentValue;
  }

  public componentDidUpdate(prevProps: Readonly<IColorSliderProps>, prevState: Readonly<IColorSliderState>): void {
    // if props changed (as opposed to a state update), set the value
    // TODO: switch to strict controlled pattern instead
    if (prevProps !== this.props && this.props.value !== undefined) {
      this.setState({ currentValue: this.props.value });
    }
  }

  public componentWillUnmount() {
    this._disposeListeners();
  }

  public render(): JSX.Element {
    const type = this._type;
    const maxValue = this._maxValue;
    const {
      // eslint-disable-next-line deprecation/deprecation
      overlayStyle,
      overlayColor,
      theme,
      className,
      styles,
      ariaLabel = type,
    } = this.props;
    const currentValue = this.value;

    const classNames = getClassNames(styles!, {
      theme: theme!,
      className,
      type,
    });

    const currentPercentage = (100 * currentValue) / maxValue;

    return (
      <div
        ref={this._root}
        className={classNames.root}
        tabIndex={0}
        onKeyDown={this._onKeyDown}
        onMouseDown={this._onMouseDown}
        role="slider"
        aria-valuenow={currentValue}
        // Narrator doesn't read aria-valuenow properly
        aria-valuetext={String(currentValue)}
        aria-valuemin={0}
        aria-valuemax={maxValue}
        aria-label={ariaLabel}
        data-is-focusable={true}
      >
        {!!(overlayColor || overlayStyle) && (
          <div
            className={classNames.sliderOverlay}
            // this isn't included in getStyles because it may change frequently
            style={
              overlayColor
                ? {
                    background:
                      type === 'transparency'
                        ? `linear-gradient(to right, #${overlayColor}, transparent)`
                        : `linear-gradient(to right, transparent, #${overlayColor})`,
                  }
                : overlayStyle
            }
          />
        )}
        <div className={classNames.sliderThumb} style={{ left: currentPercentage + '%' }} />
      </div>
    );
  }

  private get _type(): IColorSliderProps['type'] {
    // eslint-disable-next-line deprecation/deprecation
    const { isAlpha, type = isAlpha ? 'alpha' : 'hue' } = this.props;
    return type;
  }

  private get _maxValue(): number {
    return this._type === 'hue' ? MAX_COLOR_HUE : MAX_COLOR_ALPHA;
  }

  private _onKeyDown = (ev: React.KeyboardEvent): void => {
    let currentValue = this.value;
    const maxValue = this._maxValue;
    const increment = ev.shiftKey ? 10 : 1;

    // Intentionally DO NOT flip the color picker in RTL: its orientation is not very meaningful,
    // and getting all the math and styles flipped correctly is tricky
    // eslint-disable-next-line deprecation/deprecation
    switch (ev.which) {
      case KeyCodes.left: {
        currentValue -= increment;
        break;
      }
      case KeyCodes.right: {
        currentValue += increment;
        break;
      }
      case KeyCodes.home: {
        currentValue = 0;
        break;
      }
      case KeyCodes.end: {
        currentValue = maxValue;
        break;
      }
      default: {
        return;
      }
    }

    this._updateValue(ev, clamp(currentValue, maxValue));
  };

  private _onMouseDown = (ev: React.MouseEvent<HTMLElement>): void => {
    const win = getWindow(this as any);

    if (win) {
      this._disposables.push(
        on(win, 'mousemove', this._onMouseMove as (ev: MouseEvent) => void, true),
        on(win, 'mouseup', this._disposeListeners, true),
      );
    }

    this._onMouseMove(ev);
  };

  private _onMouseMove = (ev: MouseEvent | React.MouseEvent): void => {
    if (!this._root.current) {
      return;
    }

    const maxValue = this._maxValue;
    const rectSize = this._root.current.getBoundingClientRect();

    const currentPercentage = (ev.clientX - rectSize.left) / rectSize.width;
    const newValue = clamp(Math.round(currentPercentage * maxValue), maxValue);

    this._updateValue(ev, newValue);
  };

  private _disposeListeners = (): void => {
    this._disposables.forEach(dispose => dispose());
    this._disposables = [];
  };

  private _updateValue(ev: MouseEvent | KeyboardEvent | React.MouseEvent | React.KeyboardEvent, newValue: number) {
    if (newValue === this.value) {
      return;
    }

    const { onChange } = this.props;

    if (onChange) {
      onChange(ev as React.MouseEvent | React.KeyboardEvent, newValue);
    }

    if (!ev.defaultPrevented) {
      this.setState({
        currentValue: newValue,
      });
      ev.preventDefault();
    }
  }
}
