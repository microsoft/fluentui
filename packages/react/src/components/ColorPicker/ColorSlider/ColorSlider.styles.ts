import { IColorSliderStyleProps, IColorSliderStyles } from './ColorSlider.types';
import { IsFocusVisibleClassName } from '../../../Utilities';

const hueStyle = {
  background: `linear-gradient(${[
    'to left',
    'red 0',
    '#f09 10%',
    '#cd00ff 20%',
    '#3200ff 30%',
    '#06f 40%',
    '#00fffd 50%',
    '#0f6 60%',
    '#35ff00 70%',
    '#cdff00 80%',
    '#f90 90%',
    'red 100%',
  ].join(',')})`,
};

const alphaStyle = {
  backgroundImage:
    'url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAJUlEQVQYV2N89erVfwY0ICYmxoguxjgUFKI7GsTH5m4M3w1ChQC1/Ca8i2n1WgAAAABJRU5ErkJggg==)',
};

export const getStyles = (props: IColorSliderStyleProps): IColorSliderStyles => {
  // eslint-disable-next-line deprecation/deprecation
  const { theme, className, type = 'hue', isAlpha: useAlphaBackground = type !== 'hue' } = props;
  const { palette, effects } = theme;

  return {
    root: [
      'ms-ColorPicker-slider',
      {
        position: 'relative',
        height: 20,
        marginBottom: 8,
        border: `1px solid ${palette.neutralLight}`,
        borderRadius: effects.roundedCorner2,
        boxSizing: 'border-box',
        outline: 'none',

        selectors: {
          [`.${IsFocusVisibleClassName} &:focus`]: {
            outline: `1px solid ${palette.neutralSecondary}`,
          },
        },
      },
      useAlphaBackground ? alphaStyle : hueStyle,
      className,
    ],
    sliderOverlay: [
      'ms-ColorPicker-sliderOverlay',
      {
        content: '',
        position: 'absolute',
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
      },
    ],
    sliderThumb: [
      'ms-ColorPicker-thumb',
      'is-slider',
      {
        position: 'absolute',
        width: 20,
        height: 20,
        background: 'white',
        border: `1px solid ${palette.neutralSecondaryAlt}`,
        borderRadius: '50%',
        boxShadow: effects.elevation8,
        transform: 'translate(-50%, -50%)',
        top: '50%',
      },
    ],
  };
};
