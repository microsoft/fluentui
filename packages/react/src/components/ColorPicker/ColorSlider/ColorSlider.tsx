import * as React from 'react';
import { styled } from '../../../Utilities';
import { ColorSliderBase } from './ColorSlider.base';
import { getStyles } from './ColorSlider.styles';
import { IColorSliderProps, IColorSliderStyleProps, IColorSliderStyles } from './ColorSlider.types';

export const ColorSlider: React.FunctionComponent<IColorSliderProps> = styled<
  IColorSliderProps,
  IColorSliderStyleProps,
  IColorSliderStyles
>(ColorSliderBase, getStyles, undefined, { scope: 'ColorSlider' });
