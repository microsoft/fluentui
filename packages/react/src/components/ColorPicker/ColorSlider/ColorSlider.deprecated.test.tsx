import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { ColorSlider } from './ColorSlider';
import { MAX_COLOR_HUE, MAX_COLOR_ALPHA } from '../../../utilities/color/index';
import { setWarningCallback } from '../../../Utilities';

describe('ColorSlider', () => {
  let component: renderer.ReactTestRenderer | undefined;

  beforeAll(() => {
    // Prevent warn deprecations from failing test
    setWarningCallback(() => {
      /* no-op */
    });
  });

  afterAll(() => {
    setWarningCallback();
  });

  afterEach(() => {
    if (component) {
      component.unmount();
      component = undefined;
    }
  });

  it('renders hue slider correctly', () => {
    component = renderer.create(<ColorSlider value={30} minValue={0} maxValue={MAX_COLOR_HUE} ariaLabel="Hue" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders alpha slider correctly', () => {
    component = renderer.create(
      <ColorSlider
        isAlpha
        value={30}
        overlayColor="#ff0000"
        minValue={0}
        maxValue={MAX_COLOR_ALPHA}
        ariaLabel="Alpha"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
