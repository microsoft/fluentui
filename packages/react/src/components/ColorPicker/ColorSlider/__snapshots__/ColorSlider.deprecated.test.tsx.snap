// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorSlider renders alpha slider correctly 1`] = `
<div
  aria-label="Alpha"
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={30}
  aria-valuetext="30"
  className=
      ms-ColorPicker-slider
      {
        background-image: url(data;
        base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAJUlEQVQYV2N89erVfwY0ICYmxoguxjgUFKI7GsTH5m4M3w1ChQC1/Ca8i2n1WgAAAABJRU5ErkJggg==): ;
        border-radius: 2px;
        border: 1px solid #edebe9;
        box-sizing: border-box;
        height: 20px;
        margin-bottom: 8px;
        outline: none;
        position: relative;
      }
      .ms-Fabric--isFocusVisible &:focus {
        outline: 1px solid #605e5c;
      }
  data-is-focusable={true}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="slider"
  tabIndex={0}
>
  <div
    className=
        ms-ColorPicker-sliderOverlay
        {
          bottom: 0px;
          content: ;
          left: 0px;
          position: absolute;
          right: 0px;
          top: 0px;
        }
    style={
      Object {
        "background": "linear-gradient(to right, transparent, ##ff0000)",
      }
    }
  />
  <div
    className=
        ms-ColorPicker-thumb
        is-slider
        {
          background: white;
          border-radius: 50%;
          border: 1px solid #8a8886;
          box-shadow: 0 3.2px 7.2px 0 rgba(0, 0, 0, 0.132), 0 0.6px 1.8px 0 rgba(0, 0, 0, 0.108);
          height: 20px;
          position: absolute;
          top: 50%;
          transform: translate(-50%, -50%);
          width: 20px;
        }
    style={
      Object {
        "left": "30%",
      }
    }
  />
</div>
`;

exports[`ColorSlider renders hue slider correctly 1`] = `
<div
  aria-label="Hue"
  aria-valuemax={359}
  aria-valuemin={0}
  aria-valuenow={30}
  aria-valuetext="30"
  className=
      ms-ColorPicker-slider
      {
        background: linear-gradient(to left,red 0,#f09 10%,#cd00ff 20%,#3200ff 30%,#06f 40%,#00fffd 50%,#0f6 60%,#35ff00 70%,#cdff00 80%,#f90 90%,red 100%);
        border-radius: 2px;
        border: 1px solid #edebe9;
        box-sizing: border-box;
        height: 20px;
        margin-bottom: 8px;
        outline: none;
        position: relative;
      }
      .ms-Fabric--isFocusVisible &:focus {
        outline: 1px solid #605e5c;
      }
  data-is-focusable={true}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="slider"
  tabIndex={0}
>
  <div
    className=
        ms-ColorPicker-thumb
        is-slider
        {
          background: white;
          border-radius: 50%;
          border: 1px solid #8a8886;
          box-shadow: 0 3.2px 7.2px 0 rgba(0, 0, 0, 0.132), 0 0.6px 1.8px 0 rgba(0, 0, 0, 0.108);
          height: 20px;
          position: absolute;
          top: 50%;
          transform: translate(-50%, -50%);
          width: 20px;
        }
    style={
      Object {
        "left": "8.356545961002785%",
      }
    }
  />
</div>
`;
